server/auth/gitlab/gitlab.go

```diff
@@ -19,14 +24,98 @@ const (
 // Gitlab account.
 type Config struct {
 	config    *oauth2.Config
-	baseurl   string
 	group     string
 	whitelist map[string]bool
 	allusers  bool
+	apiurl    string
+	log       bool
+}
+
+// Note on Gitlab REST API calls.  We don't parse errors because it's
+// kind of a pain:
+// https://gitlab.com/help/api/README.md#data-validation-and-error-reporting
+// The two v4 api calls used are /user and /groups/:group/members/:uid
+// https://gitlab.com/help/api/users.md#for-normal-users-1
+// https://gitlab.com/help/api/members.md#get-a-member-of-a-group-or-project
+type serviceUser struct {
+	ID       int    `json:"id"`
+	Username string `json:"username"`
+	Email    string `json:"email"`
+}
+
+type serviceGroupMember struct {
+	ID          int    `json:"id"`
+	State       string `json:"state"`
+	AccessLevel int    `json:"access_level"`
+}
+
+func (c *Config) logMsg(message string) {
+	if c.log {
+		log.Print(message)
+	}
+}
+
+// A new oauth2 http client.
+func (c *Config) newClient(token *oauth2.Token) *http.Client {
+	return c.config.Client(oauth2.NoContext, token)
+}
+
+// Gets info on the current user.
+func (c *Config) getUser(token *oauth2.Token) *serviceUser {
+	client := c.newClient(token)
+	url := c.apiurl + "user"
+	resp, err := client.Get(url)
+	if err != nil {
+		return nil
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		if c.log {
+			var body bytes.Buffer
+			io.Copy(&body, resp.Body)
+			log.Printf("Gitlab error(http: %d) getting user: '%s'",
+				resp.StatusCode, body.String())
+			return nil
+		}
+	}
+	var user serviceUser
+	if err := json.NewDecoder(resp.Body).Decode(&user); err != nil {
+		return nil
+	}
+	return &user
+}
+
+// Gets current user group membership info.
+func (c *Config) checkGroupMembership(token *oauth2.Token, uid int, group string) bool {
+	client := c.newClient(token)
+	log.Printf("Checking group membership...")
+	url := fmt.Sprintf("%sgroups/%s/members/%d", c.apiurl, group, uid)
+	resp, err := client.Get(url)
+	if err != nil {
+		c.logMsg(fmt.Sprintf("Failed to get groups: %s", err))
+		return false
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		if c.log {
+			var body bytes.Buffer
```

Comment:
This is duplicated code from above. Maybe it should be a function?

Reply:
Done.